lib/spack/spack/spec.py

```diff
@@ -4732,11 +4798,25 @@ def merge_abstract_anonymous_specs(*abstract_specs: Spec):
         # Update with additional constraints from other spec
         for name in current_spec_constraint.direct_dep_difference(merged_spec):
             edge = next(iter(current_spec_constraint.edges_to_dependencies(name)))
-            merged_spec._add_dependency(edge.spec.copy(), deptypes=edge.deptypes)
+            merged_spec._add_dependency(
+                edge.spec.copy(), deptypes=edge.deptypes, virtuals=edge.virtuals
+            )
 
     return merged_spec
 
 
+def _reconstruct_virtuals_on_edges(spec):
+    """Reconstruct virtuals on edges. Used to read from old DB
+    and reindex.
+    """
+    possible_virtuals = [x for x in spec.package.dependencies if Spec(x).virtual]
+    for vspec in possible_virtuals:
+        if vspec in spec:
```

Comment:
I think you want to restrict to direct deps here?

Reply:
That should be done already, since the loop below selects only outgoing edges from the root.

Reply:
Sure, but in case vspec is not a direct dep you walk the whole dag? And what if it is a dependency of a dependency?

Reply:
Turned out the reconstruction was incomplete. See https://github.com/spack/spack/pull/35258/commits/3f1205a946536ef3cbcac332c371405023c80741